src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs

```diff
@@ -0,0 +1,361 @@
+            }
+        }
+
+        private void DragSelectionHandle(TextSelectionHandle handle)
+        {
+            if (_presenter != null && _textBox != null)
+            {
+                if (_textBox.ContextFlyout is { } flyout)
+                {
+                    flyout.Hide();
+                }
+
+                var point = ToPresenter(handle.IndicatorPosition);
+                point = point.WithY(point.Y - _presenter.FontSize / 2);
+                var hit = _presenter.TextLayout.HitTestPoint(point);
+                var caret = hit.CharacterHit.FirstCharacterIndex + hit.CharacterHit.TrailingLength;
+
+                var otherHandle = handle == _startHandle ? _endHandle : _startHandle;
+
+                if (handle.SelectionHandleType == SelectionHandleType.Start)
+                    _textBox.SelectionStart = caret;
+                else
+                    _textBox.SelectionEnd = caret;
+
+                var selectionStart = _textBox.SelectionStart;
+                var selectionEnd = _textBox.SelectionEnd;
+                var start = Math.Min(selectionStart, selectionEnd);
+                var length = Math.Max(selectionStart, selectionEnd) - start;
+
+                var rects = _presenter.TextLayout.HitTestTextRange(start, length);
+
+                if (rects.Any())
+                {
+                    var first = rects.First();
+                    var last = rects.Last();
+
+                    if (handle.SelectionHandleType == SelectionHandleType.Start)
+                        handle?.SetTopLeft(ToLayer(first.BottomLeft));
+                    else
+                        handle?.SetTopLeft(ToLayer(last.BottomRight));
+
+                    if (otherHandle.SelectionHandleType == SelectionHandleType.Start)
+                        otherHandle?.SetTopLeft(ToLayer(first.BottomLeft));
+                    else
+                        otherHandle?.SetTopLeft(ToLayer(last.BottomRight));
+                }
+
+                _presenter?.MoveCaretToTextPosition(caret);
+            }
+
+            EnsureVisible();
+        }
+
+        private Point ToLayer(Point point)
+        {
+            return (_presenter?.VisualRoot is Visual v) ? _presenter?.TranslatePoint(point, v) ?? point : point;
+        }
+
+        private Point ToPresenter(Point point)
+        {
+            return (_presenter is { } p) ? (p.VisualRoot as Visual)?.TranslatePoint(point, p) ?? point : point;
+        }
+
+        private Point ToTextBox(Point point)
+        {
+            return (_textBox is { } p) ? (p.VisualRoot as Visual)?.TranslatePoint(point, p) ?? point : point;
+        }
+
+        public void MoveHandlesToSelection()
+        {
+            if (_presenter == null || _textBox == null || _startHandle.IsDragging || _endHandle.IsDragging)
+            {
+                return;
+            }
+
+            var hasSelection = _textBox.SelectionStart != _textBox.SelectionEnd;
+            if (_caretHandle != null)
```

Comment:
The three handles are never `null` as they're all assigned only once in constructor (they can even be made readonly).